### PR TITLE
refactor: use object cache instead of event info

### PR DESCRIPTION
**Pull request by FabianKramm**
**What issue type does this pull request address?** (keep at least one, remove the others) 
/kind bugfix
/kind enhancement

**What does this pull request do? Which issues does it resolve?** (use `resolves #<issue_number>` if possible) 
resolves ENG-4997

### This PR does the following things:
* Removes the host event & delete event within syncer
* Introduces an optional object cache syncers can use to save the before state of objects, which helps to determine what exactly has changed on the other side
* Moves the patch & apply from vCluster pro into oss
* Refactors all syncers that were using the "old" bidirectional syncing approach to use the new one

### CODE REVIEW

File: pkg/syncer/syncer.go [hunk 1]
```diff
@@ -444,11 +362,14 @@ func (r *SyncController) enqueueVirtual(_ context.Context, obj client.Object, q
 	})
 }
 
-func (r *SyncController) enqueuePhysical(ctx context.Context, obj client.Object, q workqueue.TypedRateLimitingInterface[ctrl.Request], isDelete, _ bool) {
+func (r *SyncController) enqueuePhysical(ctx context.Context, obj client.Object, q workqueue.TypedRateLimitingInterface[ctrl.Request], isDelete bool) {
 	if obj == nil {
 		return
 	}
 
+	// add object to cache if it's not there yet
```
Comment (lizardruss) [inline pkg/syncer/syncer.go:370]: Comment is misleading now, and this line could be moved down next to `vReq`

**Review by zerbitx — APPROVED** [review]

File: pkg/controllers/resources/services/syncer.go [hunk 2]
```diff
@@ -117,40 +118,93 @@ func (s *serviceSyncer) Sync(ctx *synccontext.SyncContext, event *synccontext.Sy
...
 	// remove the ServiceBlockDeletion annotation if it's not needed
 	if event.Virtual.Spec.ClusterIP == event.Host.Spec.ClusterIP {
 		delete(event.Host.Annotations, ServiceBlockDeletion)
 	}
 
 	// translate selector
-	if event.Source == synccontext.SyncEventSourceHost {
-		event.Virtual.Spec.Selector = translate.VirtualLabelsMap(event.Host.Spec.Selector, event.Virtual.Spec.Selector)
-	} else {
+	// TODO: ryan - convert to bidirectional
```
Comment (zerbitx) [inline pkg/controllers/resources/services/syncer.go:203]: We can remove this todo

File: pkg/controllers/resources/priorityclasses/syncer_test.go [hunk 3]
```diff
@@ -88,9 +88,10 @@ func TestSyncToHost(t *testing.T) {
 
 			syncCtx, syncer := syncertesting.FakeStartSyncer(t, registerContext, New)
 			event := synccontext.NewSyncToHostEvent(vObj.DeepCopy())
-			if tC.isDelete {
-				event.Type = synccontext.SyncEventTypeDelete
-			}
+			// TODO: fix this
```
Comment (lizardruss) [inline pkg/controllers/resources/priorityclasses/syncer_test.go:91]: Do we need to address these TODOs or can they be removed?

**Review by lizardruss — COMMENTED** [review]
Looks like this was merged already, but some small comments

**Pull request merged**